src/main/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImpl.java

```diff
@@ -0,0 +1,909 @@
+      throws FirebaseProjectManagementException {
+    return listAndroidAppsOp(projectId).call();
+  }
+
+  @Override
+  public ApiFuture<List<AndroidApp>> listAndroidAppsAsync(String projectId) {
+    return listAndroidAppsOp(projectId).callAsync(app);
+  }
+
+  @Override
+  public List<IosApp> listIosApps(String projectId) throws FirebaseProjectManagementException {
+    return listIosAppsOp(projectId).call();
+  }
+
+  @Override
+  public ApiFuture<List<IosApp>> listIosAppsAsync(String projectId) {
+    return listIosAppsOp(projectId).callAsync(app);
+  }
+
+  private CallableOperation<List<AndroidApp>, FirebaseProjectManagementException> listAndroidAppsOp(
+      String projectId) {
+    return listAppsOp(projectId, ANDROID_APPS_RESOURCE_NAME, createAndroidAppFromAppIdFunction);
+  }
+
+  private CallableOperation<List<IosApp>, FirebaseProjectManagementException> listIosAppsOp(
+      String projectId) {
+    return listAppsOp(projectId, IOS_APPS_RESOURCE_NAME, createIosAppFromAppIdFunction);
+  }
+
+  private <T> CallableOperation<List<T>, FirebaseProjectManagementException> listAppsOp(
+      final String projectId,
+      final String platformResourceName,
+      final CreateAppFromAppIdFunction<T> createAppFromAppIdFunction) {
+    return new CallableOperation<List<T>, FirebaseProjectManagementException>() {
+      @Override
+      protected List<T> execute() throws FirebaseProjectManagementException {
+        String url = String.format(
+            "%s/v1beta1/projects/%s/%s?page_size=%d",
+            FIREBASE_PROJECT_MANAGEMENT_URL,
+            projectId,
+            platformResourceName,
+            MAXIMUM_LIST_APPS_PAGE_SIZE);
+        ImmutableList.Builder<T> builder = ImmutableList.<T>builder();
+        ListAppsResponse parsedResponse;
+        do {
+          parsedResponse = new ListAppsResponse();
+          makeGetRequest(url, parsedResponse, projectId, "Project ID");
+          if (parsedResponse.apps == null) {
+            break;
+          }
+          for (AppResponse app : parsedResponse.apps) {
+            builder.add(createAppFromAppIdFunction.apply(app.appId));
+          }
+          url = String.format(
+              "%s/v1beta1/projects/%s/%s?page_token=%s&page_size=%d",
+              FIREBASE_PROJECT_MANAGEMENT_URL,
+              projectId,
+              platformResourceName,
+              parsedResponse.nextPageToken,
+              MAXIMUM_LIST_APPS_PAGE_SIZE);
+        } while (!Strings.isNullOrEmpty(parsedResponse.nextPageToken));
+        return builder.build();
+      }
+    };
+  }
+
+  private static class ListAppsResponse {
+    @Key("apps")
+    private List<AppResponse> apps;
+
+    @Key("nextPageToken")
+    private String nextPageToken;
+  }
+
+  /* createAndroidApp, createIosApp */
+
+  @Override
+  public AndroidApp createAndroidApp(String projectId, String packageName, String displayName)
+      throws FirebaseProjectManagementException {
+    return blockOnFutureAndHandleExceptions(
```

Comment:
Ideally these blocking methods should not submit work to other threads. Can we get it to run everything on the calling thread?

```
createAndroidAppOp(projectId, packageName, displayName).call();
while (true) {
    if (isAppCreated()) {
        return result;
    }
    Thread.sleep(exponentiallyIncreasingDelay);
}
```

Reply:
Spawning threads is expensive (and even restricted) in some environments. The old GAE runtime is a good example, where there are strict caps on how many threads can be spawned by an app. Therefore we have designed our blocking methods to never spawn additional threads.

Reply:
Individual services should not spawn new threads. None of the other services (auth, fcm, iid) do, and I don't see any reason why this service should be different. Also methods like `createAndroidApp` are unlikely to be called at high volumes for performance to ever become a problem. Therefore please implement the simple/naive solution with blocking threads for now. If it ever comes to that, we can make an exception for this service and allow it to spawn threads.

Reply:
Thanks, good point. In fact, after thinking about this some more, I think more generally the only surface through which `ThreadManager` should affect threading is the asynchronous API, rather than the blocking API, so that clients that wish to have programmatic threading control should exclusively use our asynchronous API. It is indeed unnatural to force the client to choose a particular `ThreadManager` just so that their blocking calls do not spawn threads.

For this blocking version of `createAndroidApp`, I will use your original suggestion of blocking using just `Thread.sleep(...)` (will push shortly). As for my proposed asynchronous `createAndroidAppAsync` implementation, please let me know if you have any further concerns.

Reply:
My recommendation would be to keep things absolutely simple. Once you have a blocking implementation based on `Thread.sleep()`, the async version of the method can just run the blocking implementation on a thread provisioned by the `ThreadManager` -- i.e. the same way the other async methods work. We can implement something more sophisticated later, if the need ever arises.

Reply:
I will consolidate this thread with the one above. Resolving this one.